src/core/meshcop/tcat_agent.cpp

```diff
@@ -0,0 +1,524 @@
+        break;
+
+    case kTlvDecommissioning:
+        authorized = IsCommandClassAuthorized(mCommissionerAuthorizationField.mDecommissioningFlags,
+                                              mDeviceAuthorizationField.mDecommissioningFlags, nullptr);
+        break;
+
+    case kApplication:
+        authorized = IsCommandClassAuthorized(mCommissionerAuthorizationField.mApplicationFlags,
+                                              mDeviceAuthorizationField.mApplicationFlags, nullptr);
+        break;
+
+    case kInvalid:
+        authorized = false;
+        break;
+    }
+
+    return authorized;
+}
+
+TcatAgent::CommandClass TcatAgent::GetCommandClass(uint8_t aTlvType) const
+{
+    static constexpr int kGeneralTlvs            = 0x1F;
+    static constexpr int kCommissioningTlvs      = 0x3F;
+    static constexpr int kExtractionTlvs         = 0x5F;
+    static constexpr int kTlvDecommissioningTlvs = 0x7F;
+    static constexpr int kApplicationTlvs        = 0x9F;
+
+    if (aTlvType <= kGeneralTlvs)
+    {
+        return kGeneral;
+    }
+    else if (aTlvType <= kCommissioningTlvs)
+    {
+        return kCommissioning;
+    }
+    else if (aTlvType <= kExtractionTlvs)
+    {
+        return kExtraction;
+    }
+    else if (aTlvType <= kTlvDecommissioningTlvs)
+    {
+        return kTlvDecommissioning;
+    }
+    else if (aTlvType <= kApplicationTlvs)
+    {
+        return kApplication;
+    }
+    else
+    {
+        return kInvalid;
+    }
+}
+
+bool TcatAgent::CanProcessTlv(uint8_t aTlvType) const
+{
+    CommandClass tlvCommandClass = GetCommandClass(aTlvType);
+    return IsCommandClassAuthorized(tlvCommandClass);
+}
+
+Error TcatAgent::HandleSingleTlv(const Message &aIncommingMessage, Message &aOutgoingMessage)
+{
+    Error    error = kErrorParse;
+    ot::Tlv  tlv;
+    uint16_t offset = aIncommingMessage.GetOffset();
+    uint16_t length;
+    bool     response = false;
+
+    VerifyOrExit(IsConnected(), error = kErrorInvalidState);
+    SuccessOrExit(error = aIncommingMessage.Read(offset, tlv));
+
+    if (tlv.IsExtended())
+    {
+        ot::ExtendedTlv extTlv;
+        SuccessOrExit(error = aIncommingMessage.Read(offset, extTlv));
+        length = extTlv.GetLength();
+        offset += sizeof(ot::ExtendedTlv);
+    }
+    else
+    {
+        length = tlv.GetLength();
+        offset += sizeof(ot::Tlv);
+    }
+
+    if (!CanProcessTlv(tlv.GetType()))
+    {
+        error = kErrorRejected;
+    }
+    else
+    {
+        switch (tlv.GetType())
+        {
+        case kTlvDisconnect:
+            error = kErrorAbort;
+            break;
+
+        case kTlvSetActiveOperationalDataset:
+            error = HandleSetActiveOperationalDataset(aIncommingMessage, offset, length);
+            break;
+
+        case kTlvStartThreadInterface:
+            error = HandleStartThreadInterface();
+            break;
+
+        case kTlvStopThreadInterface:
+            Get<Mle::MleRouter>().Stop();
```

Comment:
It would seem cleaner here to call the general function to stop Thread (in the future, more code may be added to that function). Using: `otError otThreadSetEnabled(otInstance *aInstance, bool aEnabled)`

Reply:
fixed.